RFC-002-mapeo-protocol.md

```diff
@@ -0,0 +1,274 @@
+  platform: "desktop",
+  os: "win32",
+  arch: "x86_64",
+  size: 60241185,
+  sha256sum: "a948904f2f0f479b8f8197694b30184b0d2ed1c1cd2a1ec0fb85d299a192a447"
+}
+```
+
+Each `name` must be unique, and should self-describe the file as well as
+possible.
+
+## 4.2 `DownloadBinary(name, start=0)`
+
+This RPC is issued to begin downloading a specific application binary. The
+binary is returned as a streaming RPC response. Since binaries can be up to
+100mb or larger, an optional `start` byte offset can be sent as well, to
+facilitate peers resuming a previous download attempt that was cut off.
+
+A peer can stream this response directly to its own filesystem; likely into the
+same directory as the upgrades that it serves to others. Downloaded upgrades
+may be placed in a temporary "quarantine" directory first, so that its hash and
+digital signature can be verified, to prevent accidental retransmision of
+broken or fraudulent binaries.
+
+## 4.3 `GetPeerInfo()`
+
+Requests any metadata that the peer wishes to share. At minimum, this should
+include an object of the form
+
+```js
+{
+  plaintextProtocolVersion: "1.0.0",
+  secureProtocolVersion: "1.0.0"
+}
+```
+
+This allows the peer to see if they are able to communicate with the remote
+peer over a secure channel. If their major versions do not match, they can
+perform RPCs like the above to see if one is able to upgrade the other to a
+compatible version.
+
+An implementor should only use the major (`N.x.x`) version number. A minor
+(`x.N.x`) and patch (`x.x.N`) version are included for possible future use.
+
+## 4.4 `CreateSecureChannel()`
+
+This RPC returns a duplex stream, which first runs a NOISE handshake to
+determine a shared ephemeral key, and then wraps a XSalsa20 encryption stream
+around an inner muxrpc stream.
+
+If a secure channel already exists, further calls to this RPC must fail.
+
+# 5. Secure Channel RPCs
+
+The data sync protocol is responsible for letting peers exchange hypercore feed
+entries and media blobs. It happens over the secure channel.
+
+Requests and responses are multiplexed over the [muxrpc][] library, for
+extensibility, and so that several possible subchannels can be established,
+running various protocols:
+
+1. a `hypercore-protocol` stream
+2. a `blob-store-replication-stream` stream
+
+It's important to note that this RPC channel will be running *within* the RPC
+channel from the plaintext channel.
+
+Currently, the RPC protocol is used for coordinating the creation of hypercore
+and blob sync streams. However, it can be used for the exchange of any private
+information.
+
+One such extended use case for the RPC protocol is exchanging metadata about
+what feeds each peer has, how many entries each has, and what blobs are
+present, all before any sync operations occur. This would allow peers to
+generate a diff of what they'd upload and download if they were to sync,
+whether they have enough disk space to do so, etc.
+
+The proposed RPCs are:
```

Comment:
Is there any information that is currently shared in the initial handshake that would need to be shared before sync is requested?

Reply:
Probably just the Mapeo protocol version, which I figure we'd bump whenever we make any protocol-incompat change, like upgrading hypercore-protocol.